clang-tools-extra/clang-tidy/bugprone/CastingThroughVoidCheck.cpp

```diff
@@ -0,0 +1,50 @@
+//===--- CastingThroughVoidCheck.cpp - clang-tidy -------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "CastingThroughVoidCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "llvm/ADT/StringSet.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void CastingThroughVoidCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      explicitCastExpr(
+          hasDestinationType(qualType(pointsTo(qualType(unless(voidType()))))
+                                 .bind("target_type")),
+          hasSourceExpression(
+              explicitCastExpr(hasSourceExpression(expr(
+                                   hasType(qualType().bind("source_type")))),
+                               hasDestinationType(pointsTo(voidType())))
+                  .bind("cast")),
+          unless(builtinBitCastExpr())),
+      this);
+}
+
+void CastingThroughVoidCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto TT = *Result.Nodes.getNodeAs<QualType>("target_type");
+  const auto ST = *Result.Nodes.getNodeAs<QualType>("source_type");
+  if (Result.Context->hasSameType(TT->getUnqualifiedDesugaredType(),
```

Comment:
i think this isn't needed, even if someone cast int* to void* and back to int*, then it should be still be detected, simply because such code is stupid.